cdap-ui/app/cdap/services/datasource/index.js

```diff
@@ -0,0 +1,251 @@
+import uuid from 'node-uuid';
+import Rx from 'rx';
+
+export default class Datasource {
+  constructor() {
+    let socketData = Socket.getObservable();
+    this.bindings = {};
+
+    this.socketSubscription =  socketData.subscribe(
+      (data) => {
+        let hash = data.resource.id;
+
+        if (!this.bindings[hash]) { return; }
+
+        if (data.statusCode > 299 || data.warning) {
+          this.bindings[hash].rx.onError(data.error || data.response);
+          if (this.bindings[hash].type === 'REQUEST') {
+            this.bindings[hash].rx.onCompleted();
+            this.bindings[hash].rx.dispose();
+            delete this.bindings[hash];
+          }
+        } else {
+          this.bindings[hash].rx.onNext(data.response);
+          if (this.bindings[hash].type === 'REQUEST') {
+            this.bindings[hash].rx.onCompleted();
+            this.bindings[hash].rx.dispose();
+            delete this.bindings[hash];
+          }
+        }
+      }
+    );
+  }
+
+  request(resource = {}) {
+    let generatedResource = {
+      id: uuid.v4(),
+      json: resource.json || true,
+      method: resource.method || 'GET',
+      suppressErrors: resource.suppressErrors || false
+    };
+
+    if (resource.body) {
+      generatedResource.body = resource.body;
+    }
+    if (resource.data) {
+      generatedResource.body = resource.data;
+    }
+    if (resource.headers) {
+      generatedResource.headers = resource.headers;
+    }
+    if (resource.contentType) {
+      generatedResource.headers['Content-Type'] = resource.contentType;
+    }
+    if (!resource.url) {
+      resource.url = this.constructUrl(resource);
+    }
+
+    generatedResource.url = this.buildUrl(resource.url, resource.params);
+
+    let subject = new Rx.Subject();
+
+    this.bindings[generatedResource.id] = {
+      rx: subject,
+      resource: generatedResource,
+      type: 'REQUEST'
+    };
+
+    Socket.send({
+      action: 'request',
+      resource: generatedResource
+    });
+
+    return subject;
+  }
+
+  poll(resource = {}) {
+    let generatedResource = {
+      id: uuid.v4(),
+      interval: resource.interval || 10000,
+      json: resource.json || true,
+      method: resource.method || 'GET',
+      suppressErrors: resource.suppressErrors || false,
+    };
+
+    if (resource.body) {
+      generatedResource.body = resource.body;
+    }
+    if (resource.data) {
+      generatedResource.body = resource.data;
+    }
+    if (resource.headers) {
+      generatedResource.headers = resource.headers;
+    }
+
+    if (!resource.url) {
+      resource.url = this.constructUrl(resource);
+    }
+
+    generatedResource.url = this.buildUrl(resource.url, resource.params);
+
+    let subject = new Rx.Subject();
+
+    let observable = Rx.Observable.create((obs) => {
+      subject.subscribe(
+        (data) => {
+          obs.onNext(data);
+        },
+        (err) => {
+          obs.onError(err);
+        }
+      );
+
+      return () => {
```

Comment:
Is this supposed to be used for stopping a poll function by calling the subscriber? Based on our current example is it ```namespaceStream()``` ?